### ISSUE TITLE
Clippy fix for unwrap_or(T::default()) in the absence of type inference.

### ISSUE THREAD

**Issue by dolly-ent** [rust-lang/rust-clippy#7079]
I ran `clippy --fix` on:

```rust
fn main() {
    let s = None.unwrap_or(String::default());
    let x = s.len();
    println!("{}", x);
}
```

This fixes to

```rust
fn main() {
    let s = None.unwrap_or_default();
    let x = s.len();
    println!("{}", x);
}
```

But now we've lost the type information and it no longer compiles. It should fix to

```
fn main() {
    let s: String = None.unwrap_or_default();
    let x = s.len();
    println!("{}", x);
}
```

### Meta

- `cargo clippy -V`: clippy 0.1.51 (2fd73fab 2021-03-23)
- `rustc -Vv`:
  ```
rustc 1.51.0 (2fd73fabe 2021-03-23)
binary: rustc
commit-hash: 2fd73fabe469357a12c2c974c140f67e7cdd76d0
commit-date: 2021-03-23
host: x86_64-apple-darwin
release: 1.51.0
LLVM version: 11.0.1
  ```

**Comment by giraffate** [comment]
> But now we've lost the type information and it no longer compiles. It should fix to

Yes, or when using `None` explicitly I think it also may be better just like below:
```rust
fn main() {
    let s = String::default();
    let x = s.len();
    println!("{}", x);
}
```

**Comment by dolly-ent** [comment]
> when using `None` explicitly

The original code does not use `None` explicitly, this example is minimized.

**Comment by giraffate** [comment]
Could you provide a detailed example?

**Comment by camsteffen** [comment]
I think we should treat this as a false positive unless we can determine that the type can be inferred.

**Comment by nhamovitz** [comment]
@rustbot claim

My plan at the moment:
- if using `None` explicitly, suggest `let var = Type::default()`
- otherwise, suggest `let var: Type = expr.unwrap_or_default()`
- change applicability to something more cautious bc I don't think we're ever gonna run out of edge cases on this one

Here's some testcases I've come up with over the past couple days: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=95697b0d7d7c4bfa95f25ddca7d8747c
